File: js/pieces.js
import * as Drawing from "./drawing.js";
import { Square } from "./board.js";

"use strict";

export { Pawn };
export { Rook };
export { Knight };
export { Bishop };
export { Queen };
export { King };


function Pawn(square, white) {
    this.square = square;
    this.white = white;
    this.draw = Drawing.drawPawn;
    this.name = "pawn";

    this.copy = function () {
        return new Pawn(this.square, this.white);
    };

    this.isMoveValid = function (board, square) {
        if (!isMoveValidGeneric(this, board, square)) {
            return false;
        }
        var dir = white ? -1 : 1;
        var dx = (square.x - this.square.x);
        var dy = (square.y - this.square.y);
        if (dy * dir > 2) {
            // can never go further than two spaces
            return false;
        }
        if (dy * dir <= 0) {
            // mustn't go backwards
            return false;
        }
        if (dy * dir > 1 && this.square.y != (white ? 6 : 1)) {
            // can go two spaces if we haven't moved
            return false;
        }
        var capturedPiece = board.getPieceOnSquare(square);
        if (dx == 0 && capturedPiece != null) {
            // can't take forwards
            return false;
        }
        if (dx != 0 && (capturedPiece == null) || Math.abs(dx) > 1) {
            // can take diagonally
            return false;
        }

        return isNothingInTheWay(board, this, square);
    };

    this.getLegalMoves = function (board) {
        var moves = [];
        var dir = this.white ? -1 : 1;
        addMove(this, board, moves, this.square.offset(0, dir * 1));
        addMove(this, board, moves, this.square.offset(0, dir * 2));
        addMove(this, board, moves, this.square.offset(1, dir * 1));
        addMove(this, board, moves, this.square.offset(-1, dir * 1));

        return moves;
    };

}

function Rook(square, white) {
    this.square = square;
    this.white = white;
    this.draw = Drawing.drawRook;
    this.name = "rook";

    this.copy = function () {
        return new Rook(this.square, this.white);
    };

    this.getLegalMoves = function (board) {
        var moves = [];
        var x = this.square.x;
        var y = this.square.y;
        while (addMove(this, board, moves, new Square(++x, y)));
        x = this.square.x;
        y = this.square.y;
        while (addMove(this, board, moves, new Square(--x, y)));
        x = this.square.x;
        y = this.square.y;
        while (addMove(this, board, moves, new Square(x, ++y)));
        x = this.square.x;
        y = this.square.y;
        while (addMove(this, board, moves, new Square(x, --y)));

        return moves;
    };
}

function inBounds(square) {
    return (square.x >= 0 && square.x <= 7 && square.y >= 0 && square.y <= 7);
}

function addMove(piece, board, moves, square) {
    if (square.equals(piece.square)) {
        return true;
    }
    if (inBounds(square) && piece.isMoveValid(board, square)) {
        moves.push(square);
        return true;
    }
    return false;
}

function Knight(square, white) {
    this.square = square;
    this.white = white;
    this.draw = Drawing.drawKnight;
    this.name = "Knight";

    this.copy = function () {
        return new Knight(this.square, this.white);
    };

    this.getLegalMoves = function (board) {
        return [];
    };

}

function Bishop(square, white) {
    this.square = square;
    this.white = white;
    this.draw = Drawing.drawBishop;
    this.name = "Bishop";
    this.copy = function () {
        return new Bishop(this.square, this.white);
    };
    this.getLegalMoves = function (board) {
        return [];
    };

}

function Queen(square, white) {
    this.square = square;
    this.white = white;
    this.draw = Drawing.drawQueen;
    this.name = "Queen";

    this.copy = function () {
        return new Queen(this.square, this.white);
    };

    this.getLegalMoves = function (board) {
        return [];
    };
}

function King(square, white) {
    this.square = square;
    this.white = white;
    this.draw = Drawing.drawKing;
    this.name = "King";

    this.copy = function () {
        return new King(this.square, this.white);
    };

    this.getLegalMoves = function (board) {
        return [];
    };

}

function isMoveValidGeneric(piece, board, square) {
    var valid = inBounds(square);
    var pieceCaptured = board.getPieceOnSquare(square);
    if (pieceCaptured != null) {
        if (pieceCaptured.white == piece.white) {
            // can't capture our own guys
            valid = false;
        }
    }
    return valid;
}

function isNothingInTheWay(board, piece, square) {
    var dx = square.x - piece.square.x;
    var dy = square.y - piece.square.y;
    var xinc = (dx == 0) ? 0 : ((dx < 0) ? -1 : 1);  // ie. either no horizontal move or left / right
    var yinc = (dy == 0) ? 0 : ((dy < 0) ? -1 : 1);  // ie. either no vertical move or up / down

    if (xinc == 0 && yinc == 0) {
        throw ("something's way wrong here");
    }
    var s = piece.square;
    var foundSomeone = false;
    do {    // walk in the specified direction until we get there or we find a piece
        s = s.offset(xinc, yinc);
        if (s.equals(square)) {
            return true;
        }
        if (board.getPieceOnSquare(s)) {
            foundSomeone = true;
        }
    } while (!foundSomeone);

    return false;
}

File: ChessDotNetBackend/Pieces.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace ChessDotNetBackend
{
    public enum PieceType
    {
        WhitePawn,
        WhiteRook,
        WhiteKnight,
        WhiteBishop,
        WhiteQueen,
        WhiteKing,
        BlackPawn,
        BlackRook,
        BlackKnight,
        BlackBishop,
        BlackQueen,
        BlackKing
    }

    public interface IPieceVisitor
    {
        void Visit(Pawn pawn, object data);
        void Visit(Rook rook, object data);
        void Visit(Knight knight, object data);
        void Visit(Bishop bishop, object data);
        void Visit(Queen queen, object data);
        void Visit(King king, object data);
    }

    public interface IPiece
    {
        Square CurrentPosition { get; set; }
        bool White { get; }
        string Name { get; }

        void Accept(IPieceVisitor visitor, object data);
        IPiece Copy();
        IEnumerable<Square> GetAllMoves(Board m_board);
        bool IsMoveValid(Board board, Square destination);
        double Value { get; }
        PieceType PieceType { get; }
        bool IsPawn { get; }
    }

    static class Piece
    {
        public static bool addMove(IPiece piece, Board board, List<Square> moves, Square destination)
        {
            if (destination == piece.CurrentPosition)
            {
                return true;
            }
            if (piece.IsMoveValid(board, destination))
            {
                moves.Add(destination);
                return true;
            }
            return false;
        }

        public static bool IsMoveValid(IPiece piece, Board board, Square destination)
        {
            if (!destination.InBounds)
            {
                return false;
            }
            IPiece capturedPiece = board.GetPieceOnSquare(destination);
            bool correctCapture = capturedPiece == null || capturedPiece.White != piece.White;
            return correctCapture;
        }
    }

    public class Pawn : IPiece
    {
        public Square CurrentPosition { get; set; }
        public bool White { get; }
        public string Name { get; }

        public Pawn(Square currentPosition, bool white)
        {
            CurrentPosition = currentPosition;
            White = white;
            Name = White ? "White Pawn" : "Black Pawn";
            PieceType = white ? PieceType.WhitePawn : PieceType.BlackPawn;
        }

        public void Accept(IPieceVisitor visitor, object data) => visitor.Visit(this, data);

        public IEnumerable<Square> GetAllMoves(Board board)
        {
            var moves = new List<Square>();
            var dir = White ? -1 : 1;
            Piece.addMove(this, board, moves, CurrentPosition.Offset(0, dir * 1));
            Piece.addMove(this, board, moves, CurrentPosition.Offset(0, dir * 2));
            Piece.addMove(this, board, moves, CurrentPosition.Offset(1, dir * 1));
            Piece.addMove(this, board, moves, CurrentPosition.Offset(-1, dir * 1));
            return moves;
        }

        public bool IsMoveValid(Board board, Square destination)
        {
            if (!Piece.IsMoveValid(this, board, destination))
            {
                return false;
            }
            int nDir = White ? -1 : 1;  // which way can this pawn move?
            if ((destination.y - CurrentPosition.y) * nDir > 2)   // can never go further than two spaces
                return false;
            if ((destination.y - CurrentPosition.y) * nDir <= 0)  // mustn't go backwards
                return false;
            if (((destination.y - CurrentPosition.y) * nDir > 1) && (CurrentPosition.y != (White ? 6 : 1)))  // can go two spaces if we haven't moved
                return false;
            if ((destination.x == CurrentPosition.x) && board.GetPieceOnSquare(destination) != null)  // can't take forwards
                return false;
            if (destination.x != CurrentPosition.x)
            {
                if (board.GetPieceOnSquare(destination) == null)
                {
                    return false;
                }
                if (Math.Abs(destination.x - CurrentPosition.x) > 1)  // can take diagonally
                {
                    return false;
                }
            }
            return board.IsNothingInTheWay(this, destination);

        }

        public override string ToString() => Name + " " + CurrentPosition.ToString();

        public IPiece Copy() => new Pawn(CurrentPosition, White);

        public override bool Equals(object obj)
        {
            if (obj == null || GetType() != obj.GetType())
            {
                return false;
            }
            Pawn p = (Pawn)obj;
            return p.White == White && p.CurrentPosition == CurrentPosition;
        }

        public double Value => 1;

        public PieceType PieceType { get; }

        public bool IsPawn => true;
    }

    public class Rook : IPiece
    {
        public Square CurrentPosition { get; set; }
        public bool White { get; }
        public string Name { get; }

        public Rook(Square currentPosition, bool white)
        {
            CurrentPosition = currentPosition;
            White = white;
            Name = White ? "White Rook" : "Black Rook";
            PieceType = white ? PieceType.WhiteRook : PieceType.BlackRook;
        }

        public void Accept(IPieceVisitor visitor, object data) => visitor.Visit(this, data);

        public IEnumerable<Square> GetAllMoves(Board board)
        {
            var moves = new List<Square>();
            int x = CurrentPosition.x;
            int y = CurrentPosition.y;
            while (Piece.addMove(this, board, moves, new Square(x++, y)))
            { }
            x = CurrentPosition.x;
            y = CurrentPosition.y;
            while (Piece.addMove(this, board, moves, new Square(x--, y)))
            { }
            x = CurrentPosition.x;
            y = CurrentPosition.y;
            while (Piece.addMove(this, board, moves, new Square(x, y++)))
            { }
            x = CurrentPosition.x;
            y = CurrentPosition.y;
            while (Piece.addMove(this, board, moves, new Square(x, y--)))
            { }
            return moves;
        }

        public bool IsMoveValid(Board board, Square destination)
        {
            if (!Piece.IsMoveValid(this, board, destination))
            {
                return false;
            }
            if (CurrentPosition.x != destination.x && CurrentPosition.y != destination.y)
            {
                return false;
            }
            return board.IsNothingInTheWay(this, destination);
        }

        public override string ToString() => Name + " " + CurrentPosition.ToString();

        public IPiece Copy() => new Rook(CurrentPosition, White);

        public override bool Equals(object obj)
        {
            if (obj == null || GetType() != obj.GetType())
            {
                return false;
            }
            Rook p = (Rook)obj;
            return p.White == White && p.CurrentPosition == CurrentPosition;
        }

        public double Value => 5;

        public PieceType PieceType { get; }

        public bool IsPawn => false;
    }

    public class Knight : IPiece
    {
        public Square CurrentPosition { get; set; }
        public bool White { get; }
        public string Name { get; }

        public Knight(Square currentPosition, bool white)
        {
            CurrentPosition = currentPosition;
            White = white;
            Name = White ? "White Knight" : "Black Knight";
            PieceType = white ? PieceType.WhiteKnight : PieceType.BlackKnight;
        }

        public void Accept(IPieceVisitor visitor, object data) => visitor.Visit(this, data);

        public IEnumerable<Square> GetAllMoves(Board board)
        {
            var moves = new List<Square>();
            Piece.addMove(this, board, moves, CurrentPosition.Offset(2, 1));
            Piece.addMove(this, board, moves, CurrentPosition.Offset(2, -1));
            Piece.addMove(this, board, moves, CurrentPosition.Offset(-2, 1));
            Piece.addMove(this, board, moves, CurrentPosition.Offset(-2, -1));
            Piece.addMove(this, board, moves, CurrentPosition.Offset(1, 2));
            Piece.addMove(this, board, moves, CurrentPosition.Offset(-1, 2));
            Piece.addMove(this, board, moves, CurrentPosition.Offset(1, -2));
            Piece.addMove(this, board, moves, CurrentPosition.Offset(-1, -2));
            return moves;
        }

        public bool IsMoveValid(Board board, Square destination)
        {
            if (!Piece.IsMoveValid(this, board, destination))
            {
                return false;
            }
            var dx = destination.x - CurrentPosition.x;
            var dy = destination.y - CurrentPosition.y;
            //  two spaces forward, one to either side...
            return (Math.Abs(dy) == 2 && Math.Abs(dx) == 1) || (Math.Abs(dy) == 1 && Math.Abs(dx) == 2);
        }

        public override string ToString() => Name + " " + CurrentPosition.ToString();

        public IPiece Copy() => new Knight(CurrentPosition, White);

        public override bool Equals(object obj)
        {
            if (obj == null || GetType() != obj.GetType())
            {
                return false;
            }
            Knight p = (Knight)obj;
            return p.White == White && p.CurrentPosition == CurrentPosition;
        }

        public double Value => 3;

        public PieceType PieceType { get; }

        public bool IsPawn => false;
    }

    public class Bishop : IPiece
    {
        public Square CurrentPosition { get; set; }
        public bool White { get; }
        public string Name { get; }

        public Bishop(Square currentPosition, bool white)
        {
            CurrentPosition = currentPosition;
            White = white;
            Name = White ? "White Bishop" : "Black Bishop";
            PieceType = white ? PieceType.WhiteBishop: PieceType.BlackBishop;
        }

        public void Accept(IPieceVisitor visitor, object data) => visitor.Visit(this, data);

        public IEnumerable<Square> GetAllMoves(Board board)
        {
            var moves = new List<Square>();
            int x = CurrentPosition.x;
            int y = CurrentPosition.y;
            while (Piece.addMove(this, board, moves, new Square(x++, y++)))
                ;
            x = CurrentPosition.x;
            y = CurrentPosition.y;
            while (Piece.addMove(this, board, moves, new Square(x--, y++)))
                ;
            x = CurrentPosition.x;
            y = CurrentPosition.y;
            while (Piece.addMove(this, board, moves, new Square(x--, y--)))
                ;
            x = CurrentPosition.x;
            y = CurrentPosition.y;
            while (Piece.addMove(this, board, moves, new Square(x++, y--))) ;
            return moves;
        }

        public bool IsMoveValid(Board board, Square destination)
        {
            if (!Piece.IsMoveValid(this, board, destination))
            {
                return false;
            }
            if (Math.Abs(destination.x - CurrentPosition.x) != Math.Abs(destination.y - CurrentPosition.y))
            {
                return false;
            }
            return board.IsNothingInTheWay(this, destination);
        }

        public override string ToString() => Name + " " + CurrentPosition.ToString();

        public IPiece Copy() => new Bishop(CurrentPosition, White);

        public override bool Equals(object obj)
        {
            if (obj == null || GetType() != obj.GetType())
            {
                return false;
            }
            Bishop p = (Bishop)obj;
            return p.White == White && p.CurrentPosition == CurrentPosition;
        }

        public double Value => 3;

        public PieceType PieceType { get; }

        public bool IsPawn => false;
    }

    public class Queen : IPiece
    {
        public Square CurrentPosition { get; set; }
        public bool White { get; }
        public string Name { get; }

        public Queen(Square currentPosition, bool white)
        {
            CurrentPosition = currentPosition;
            White = white;
            Name = White ? "White Queen" : "Black Queen";
            PieceType = white ? PieceType.WhiteQueen : PieceType.BlackQueen;
        }

        public void Accept(IPieceVisitor visitor, object data) => visitor.Visit(this, data);

        public IEnumerable<Square> GetAllMoves(Board board)
        {
            var moves = new List<Square>();
            int x = CurrentPosition.x;
            int y = CurrentPosition.y;
            while (Piece.addMove(this, board, moves, new Square(x++, y++)))
                ;
            x = CurrentPosition.x;
            y = CurrentPosition.y;
            while (Piece.addMove(this, board, moves, new Square(x--, y++)))
                ;
            x = CurrentPosition.x;
            y = CurrentPosition.y;
            while (Piece.addMove(this, board, moves, new Square(x--, y--)))
                ;
            x = CurrentPosition.x;
            y = CurrentPosition.y;
            while (Piece.addMove(this, board, moves, new Square(x++, y--)))
                ;
            x = CurrentPosition.x;
            y = CurrentPosition.y;
            while (Piece.addMove(this, board, moves, new Square(x++, y)))
                ;
            x = CurrentPosition.x;
            y = CurrentPosition.y;
            while (Piece.addMove(this, board, moves, new Square(x--, y)))
                ;
            x = CurrentPosition.x;
            y = CurrentPosition.y;
            while (Piece.addMove(this, board, moves, new Square(x, y++)))
                ;
            x = CurrentPosition.x;
            y = CurrentPosition.y;
            while (Piece.addMove(this, board, moves, new Square(x, y--)))
                ;
            return moves;
        }

        public bool IsMoveValid(Board board, Square destination)
        {
            if (!Piece.IsMoveValid(this, board, destination))
            {
                return false;
            }
            if (CurrentPosition.x != destination.x && CurrentPosition.y != destination.y)
            {
                if (Math.Abs(destination.x - CurrentPosition.x) != Math.Abs(destination.y - CurrentPosition.y))
                {
                    return false;
                }
            }
            return board.IsNothingInTheWay(this, destination);
        }

        public override string ToString() => Name + " " + CurrentPosition.ToString();

        public IPiece Copy() => new Queen(CurrentPosition, White);

        public override bool Equals(object obj)
        {
            if (obj == null || GetType() != obj.GetType())
            {
                return false;
            }
            Queen p = (Queen)obj;
            return p.White == White && p.CurrentPosition == CurrentPosition;
        }

        public double Value => 9;

        public PieceType PieceType { get; }

        public bool IsPawn => false;
    }

    public class King : IPiece
    {
        public Square CurrentPosition { get; set; }
        public bool White { get; }
        public string Name { get; }

        public King(Square currentPosition, bool white)
        {
            CurrentPosition = currentPosition;
            White = white;
            Name = White ? "White King" : "Black King";
            PieceType = white ? PieceType.WhiteKing : PieceType.BlackKing;
        }

        public void Accept(IPieceVisitor visitor, object data) => visitor.Visit(this, data);

        public IEnumerable<Square> GetAllMoves(Board board)
        {
            var moves = new List<Square>();
            Piece.addMove(this, board, moves, CurrentPosition.Offset(0, 1));
            Piece.addMove(this, board, moves, CurrentPosition.Offset(0, -1));
            Piece.addMove(this, board, moves, CurrentPosition.Offset(1, 1));
            Piece.addMove(this, board, moves, CurrentPosition.Offset(1, 0));
            Piece.addMove(this, board, moves, CurrentPosition.Offset(1, -1));
            Piece.addMove(this, board, moves, CurrentPosition.Offset(-1, 1));
            Piece.addMove(this, board, moves, CurrentPosition.Offset(-1, 0));
            Piece.addMove(this, board, moves, CurrentPosition.Offset(-1, -1));
            return moves;
        }

        public bool IsMoveValid(Board board, Square destination)
        {
            if (!Piece.IsMoveValid(this, board, destination))
            {
                return false;
            }
            if (Math.Abs(destination.x - CurrentPosition.x) > 1)
            {
                return false;
            }
            if (Math.Abs(destination.y - CurrentPosition.y) > 1)
            {
                return false;
            }
            return true;
        }

        public override string ToString() => Name + " " + CurrentPosition.ToString();

        public IPiece Copy() => new King(CurrentPosition, White);

        public override bool Equals(object obj)
        {
            if (obj == null || GetType() != obj.GetType())
            {
                return false;
            }
            King p = (King)obj;
            return p.White == White && p.CurrentPosition == CurrentPosition;
        }

        public double Value => 1000;

        public PieceType PieceType { get; }

        public bool IsPawn => false;
    }

}
